### PR TITLE
API: Add 'preprocessCmd' function to RaftStateMachine typeclass

### PR DESCRIPTION
**Pull request by tdietert**

### CODE REVIEW

File: src/Raft/StateMachine.hs [hunk 1]
```diff
@@ -1,7 +1,12 @@
 {-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE ScopedTypeVariables #-}
 {-# LANGUAGE FunctionalDependencies #-}
 {-# LANGUAGE TypeFamilies #-}
 {-# LANGUAGE TypeFamilyDependencies #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE DefaultSignatures #-}
+{-# LANGUAGE AllowAmbiguousTypes #-}
```
Comment (faineance) [inline src/Raft/StateMachine.hs]: With this, and the usage in applyLogCmd, will AllowAmbiguousTypes be needed when using the API that's surfaced?

Comment (tdietert) [reply src/Raft/StateMachine.hs]: No, `applyLogCmd` is used internally and the raft library users will not have to ever call these functions themselves; thus, the ambiguity is resolved using type applications, with the core Raft library event handling loop.

Comment (tdietert) [reply src/Raft/StateMachine.hs]: see https://github.com/adjoint-io/raft/pull/119/files#diff-6a654c25ca3c82a99a2cdd744040e4f0R90

Comment (faineance) [reply src/Raft/StateMachine.hs]: Ok Cool

**Pull request merged**